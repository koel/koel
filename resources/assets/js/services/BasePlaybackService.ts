import isMobile from 'ismobilejs'
import { throttle } from 'lodash'
import plyr from 'plyr'
import { watch } from 'vue'
import { volumeManager } from '@/services/volumeManager'

export abstract class BasePlaybackService {
  public player!: Plyr
  private boundMediaEvents = new Set<[string, EventListener, boolean]>()

  public activate (plyrWrapper: HTMLElement) {
    if (!this.player) {
      // player and watchers can be kept between init/destroy sessions
      this.player = this.player || plyr.setup(plyrWrapper, { controls: [] })[0]
      watch(volumeManager.volume, volume => this.player.setVolume(volume), { immediate: true })
      this.setMediaSessionActionHandlers()
    }

    if (!this.boundMediaEvents.size) {
      this.addMediaEventListeners()
    }

    return this
  }

  protected setMediaSessionActionHandlers () {
    if (!navigator.mediaSession) {
      return
    }

    navigator.mediaSession.setActionHandler('play', () => this.resume())
    navigator.mediaSession.setActionHandler('pause', () => this.pause())
    navigator.mediaSession.setActionHandler('stop', () => this.stop())
    navigator.mediaSession.setActionHandler('previoustrack', () => this.playPrev())
    navigator.mediaSession.setActionHandler('nexttrack', () => this.playNext())

    if (!isMobile.apple) {
      navigator.mediaSession.setActionHandler('seekbackward', details => this.rewind(details.seekOffset || 10))
      navigator.mediaSession.setActionHandler('seekforward', details => this.forward(details.seekOffset || 10))
    }

    navigator.mediaSession.setActionHandler('seekto', details => {
      if (details.fastSeek && 'fastSeek' in this.player.media) {
        this.fastSeek(details.seekTime || 0)
      } else {
        this.seekTo(details.seekTime || 0)
      }
    })
  }

  private addMediaEventListeners () {
    const listen = (event: string, handler: Closure, options?: boolean | AddEventListenerOptions) => {
      this.player.media.addEventListener(event, handler, options)
      this.boundMediaEvents.add([event, handler, !!options])
    }

    listen('error', this.onError.bind(this), true)
    listen('ended', this.onEnded.bind(this))

    const timeUpdateHandler = process.env.NODE_ENV === 'test' ? this.onTimeUpdate : throttle(this.onTimeUpdate, 1000)
    listen('timeupdate', timeUpdateHandler.bind(this))
  }

  public abstract play (source: Streamable): Promise<void>

  public abstract stop (): Promise<void>

  public abstract pause (): Promise<void>

  public abstract resume (): Promise<void>

  public abstract playNext (): Promise<void>

  public abstract playPrev (): Promise<void>

  public abstract toggle (): Promise<void>

  public abstract rewind (seconds: number): void

  public abstract forward (seconds: number): void

  public abstract seekTo (position: number): void

  public abstract fastSeek (position: number): void

  protected abstract onError (error: ErrorEvent): void

  protected abstract onEnded (event: Event): void

  protected abstract onTimeUpdate (event: Event): void

  public abstract rotateRepeatMode (): void

  public deactivate () {
    // Upon deactivating (i.e. switching to another playback service) we want to remove all event listeners
    // to not mess up the media events in the other service.
    this.boundMediaEvents.forEach(([event, handler, options]) => {
      this.player.media.removeEventListener(event, handler, options)
    })

    this.boundMediaEvents.clear()

    this.stop()
  }
}
