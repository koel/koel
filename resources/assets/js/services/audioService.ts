import { equalizerStore } from '@/stores'
import { frequencies } from '@/config'
import { dbToGain } from '@/utils'

interface Band {
  label: string
  filter: BiquadFilterNode
  db: number
}

export const audioService = {
  unlocked: false,

  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,
  preampGainNode: null as unknown as GainNode,
  analyzer: null as unknown as AnalyserNode,

  bands: [] as Band[],

  init (mediaElement: HTMLMediaElement) {
    this.element = mediaElement

    this.context = new AudioContext()
    this.preampGainNode = this.context.createGain()
    this.source = this.context.createMediaElementSource(this.element)
    this.analyzer = this.context.createAnalyser()

    this.source.connect(this.preampGainNode)

    const config = equalizerStore.getConfig()

    this.changePreampGain(config.preamp)

    let prevFilter: BiquadFilterNode

    // Create 10 bands with the frequencies similar to those of Winamp and connect them together.
    frequencies.forEach((frequency, i) => {
      const filter = this.context.createBiquadFilter()

      if (i === 0) {
        filter.type = 'lowshelf'
      } else if (i === frequencies.length - 1) {
        filter.type = 'highshelf'
      } else {
        filter.type = 'peaking'
      }

      filter.Q.setTargetAtTime(1, this.context.currentTime, 0.01)
      filter.frequency.setTargetAtTime(frequency, this.context.currentTime, 0.01)
      filter.gain.value = dbToGain(config.gains[i])

      prevFilter ? prevFilter.connect(filter) : this.preampGainNode.connect(filter)
      prevFilter = filter

      this.bands.push({
        filter,
        label: String(frequency).replace('000', 'K'),
        db: config.gains[i]
      })
    })

    prevFilter!.connect(this.analyzer)

    // connect the analyzer node last, so that changes to the equalizer affect the visualizer as well
    this.analyzer.connect(this.context.destination)

    this.unlockAudioContext()
  },

  changePreampGain (db: number) {
    this.preampGainNode.gain.value = dbToGain(db)
  },

  changeFilterGain (node: BiquadFilterNode, db: number) {
    this.bands.find(band => band.filter === node)!.db = db
    node.gain.value = dbToGain(db)
  },

  /**
   * Attempt to unlock the audio context on mobile devices by creating and playing a silent buffer upon the
   * first user interaction.
   */
  unlockAudioContext () {
    ['touchend', 'touchstart', 'click'].forEach(event => {
      document.addEventListener(event, () => {
        if (this.unlocked) return

        const source = this.context.createBufferSource()
        source.buffer = this.context.createBuffer(1, 1, 22050)
        source.connect(this.context.destination)
        source.start(0)

        this.unlocked = true
      }, {
        once: true
      })
    })
  }
}
