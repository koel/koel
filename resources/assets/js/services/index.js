export * from "./info";
export * from "./download";
export * from "./http";
export * from "./ls";
export * from "./playback";
export * from "./youtube";
export * from "./socket";
