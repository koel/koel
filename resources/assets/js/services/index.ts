export * from './http'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
export * from './mediaInfoService'
export * from './cache'
export * from './socketListener'
export * from './volumeManager'
