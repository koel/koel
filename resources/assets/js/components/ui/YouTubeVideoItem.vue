<template>
  <a :href="url" data-testid="youtube-search-result" role="button" @click.prevent="play">
    <img :alt="video.snippet.title" :src="video.snippet.thumbnails.default.url" width="90">
    <div class="meta">
      <h3 class="title">{{ video.snippet.title }}</h3>
      <p class="desc">{{ video.snippet.description }}</p>
    </div>
  </a>
</template>

<script lang="ts" setup>
import { computed, toRefs } from 'vue'
import { youTubeService } from '@/services'
import { RouterKey } from '@/symbols'
import { requireInjection } from '@/utils'

const router = requireInjection(RouterKey)

const props = defineProps<{ video: YouTubeVideo }>()
const { video } = toRefs(props)

const url = computed(() => `https://youtu.be/${video.value.id.videoId}`)

const play = () => {
  youTubeService.play(video.value)
  router.go('youtube')
}
</script>

<style lang="scss" scoped>
a {
  display: flex;
  gap: 10px;

  &:hover, &:active, &:focus {
    color: var(--color-text-primary);
  }
}

.title {
  font-size: 1.1rem;
  margin-bottom: .4rem;
}

.desc {
  font-size: .9rem;
}

img {
  align-self: self-start;
}

.meta {
  overflow-wrap: anywhere;
}
</style>
