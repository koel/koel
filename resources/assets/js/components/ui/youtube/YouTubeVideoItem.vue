<template>
  <a
    :href="url"
    class="flex gap-3 !text-k-text-secondary hover:!text-k-text-primary focus:!text-k-text-primary active:!text-k-text-primary"
    data-testid="youtube-search-result"
    role="button"
    @click.prevent="play"
  >
    <img :alt="video.snippet.title" :src="video.snippet.thumbnails.default.url" class="self-start" width="90">
    <aside class="space-y-1">
      <h3 class="text-lg">{{ unescape(video.snippet.title) }}</h3>
      <p class="text-[0.9rem]">{{ video.snippet.description }}</p>
    </aside>
  </a>
</template>

<script lang="ts" setup>
import { unescape } from 'lodash'
import { computed, toRefs } from 'vue'
import { youTubeService } from '@/services'
import { useRouter } from '@/composables'

const props = defineProps<{ video: YouTubeVideo }>()

const { go } = useRouter()

const { video } = toRefs(props)

const url = computed(() => `https://youtu.be/${video.value.id.videoId}`)

const play = () => {
  youTubeService.play(video.value)
  go('youtube')
}
</script>

<style lang="postcss" scoped>
aside {
  overflow-wrap: anywhere;
}
</style>
