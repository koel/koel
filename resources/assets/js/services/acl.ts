import { http } from '@/services/http'
import { cache } from '@/services/cache'

type ResourceType = 'album' | 'artist' | 'radio-station' | 'user'
type Action = 'edit' | 'delete'

export const acl = {
  checkResourcePermission: async (type: ResourceType, id: string | number, action: Action) => {
    return await cache.remember(['permission', type, id, action], async () => {
      const { allowed } = await http.silently.get<{ allowed: boolean }>(`acl/permissions/${type}/${id}/${action}`)

      return allowed
    })
  },

  fetchAssignableRoles: async () => {
    return await cache.remember(['assignable-roles'], async () => {
      const { roles } = await http.get<{
        roles: Array<{ id: Role, label: string, description: string }>
      }>('acl/assignable-roles')

      return roles
    })
  },
}
