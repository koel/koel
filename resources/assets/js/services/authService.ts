import { localStorageService } from '@/services'

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
