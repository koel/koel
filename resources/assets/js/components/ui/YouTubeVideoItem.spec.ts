import { expect, it } from 'vitest'
import { screen } from '@testing-library/vue'
import { youTubeService } from '@/services'
import UnitTestCase from '@/__tests__/UnitTestCase'
import YouTubeVideoItem from './YouTubeVideoItem.vue'

let video: YouTubeVideo

new class extends UnitTestCase {
  private renderComponent () {
    video = {
      id: {
        videoId: 'cLgJQ8Zj3AA'
      },
      snippet: {
        title: 'Guess what it is',
        description: 'From the LA Opening Gala 2014: John Williams Celebration',
        thumbnails: {
          default: {
            url: 'https://i.ytimg.com/an_webp/cLgJQ8Zj3AA/mqdefault_6s.webp'
          }
        }
      }
    }
    return this.render(YouTubeVideoItem, {
      props: {
        video
      }
    })
  }

  protected test () {
    it('renders', () => expect(this.renderComponent().html()).toMatchSnapshot())

    it('plays', async () => {
      const mock = this.mock(youTubeService, 'play')
      this.renderComponent()

      await this.user.click(screen.getByRole('button'))

      expect(mock).toHaveBeenCalledWith(video)
    })
  }
}
