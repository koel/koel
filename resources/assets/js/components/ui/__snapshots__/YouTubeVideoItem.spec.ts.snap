// Vitest Snapshot v1

exports[`renders 1`] = `
<a href="https://youtu.be/cLgJQ8Zj3AA" data-testid="youtube-search-result" role="button" data-v-fd1eda6f=""><img alt="Guess what it is" src="https://i.ytimg.com/an_webp/cLgJQ8Zj3AA/mqdefault_6s.webp" width="90" data-v-fd1eda6f="">
  <div class="meta" data-v-fd1eda6f="">
    <h3 class="title" data-v-fd1eda6f="">Guess what it is</h3>
    <p class="desc" data-v-fd1eda6f="">From the LA Opening Gala 2014: John Williams Celebration</p>
  </div>
</a>
`;
